[Docs] New Drivers and ORMs Section (#10878)
yugabyte/yugabyte-db: docs/content/latest/drivers-orms/go/pq.md

---
title: Go Drivers
linkTitle: Go Drivers
description: Go Drivers for YSQL
headcontent: Go Drivers for YSQL
image: /images/section_icons/sample-data/s_s1-sampledata-3x.png
menu:
  latest:
    name: Go Drivers
    identifier: pq-driver
    parent: go-drivers
    weight: 400
isTocNested: true
showAsideToc: true
---

For Go Applications, most drivers provide database connectivity through the standard `database/sql`
API. YugabyteDB supports [PGX Driver](https://github.com/jackc/pgx) and the
[PQ Driver](https://github.com/lib/pq).

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/drivers-orms/go/pgx/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      PGX Driver
    </a>
  </li>

  <li >
    <a href="/latest/drivers-orms/go/pq/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      PQ Driver
    </a>
  </li>

</ul>

The [PQ driver](https://github.com/lib/pq/) is a popular driver for PostgreSQL which can used for connecting to YugabyteDB YSQL as well.

This driver allows Go programmers to connect to YugabyteDB database to execute DMLs and DDLs using
the standard `database/sql` package.

> ## Quick start

Author: Made the required changes.

This driver allows Go programmers to connect to YugabyteDB to execute DMLs and DDLs using
the standard `database/sql` package.

## CRUD operations with PQ driver

Learn how to establish a connection to YugabyteDB database and begin simple CRUD operations using
the steps in the [Build an application](../../../quick-start/build-apps/go/ysql-pq) page under the
Quick start section.

The following sections break down the quick start example to demonstrate how to perform common tasks required for Go application development using the PQ driver.

### Step 1: Import the driver package

Import the PQ driver package by adding the following import statement in your Go code.

```go
import (
  _ "github.com/lib/pq"
)
```

### Step 2: Connect to YugabyteDB database

Go applications can connect to YugabyteDB using the `sql.Open()` function.
The `sql` package includes all the functions or structs required for working with YugabyteDB.

Use the `sql.Open()` function to create a connection object for the YugabyteDB database. This can be
used for performing DDLs and DMLs against the database.

> The connection details can be specified either as string params or via an url in the format given below:

Reviewer: ```suggestion
The connection details can be specified either as string parameters or via a URL in the following format:
```

```go
postgresql://username:password@hostname:port/database
```

Code snippet for connecting to YugabyteDB:

```go
psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s",
                        host, port, user, password, dbname)
// Other connection configs are read from the standard environment variables:
// PGSSLMODE, PGSSLROOTCERT, and so on.
db, err := sql.Open("postgres", psqlInfo)
defer db.Close()
if err != nil {
    log.Fatal(err)
}
```

| Parameter | Description | Default |
| :---------- | :---------- | :------ |
| host  | hostname of the YugabyteDB instance | localhost
| port |  Listen port for YSQL | 5433
| user | user for connecting to the database | yugabyte
| password | password for connecting to the database | yugabyte
| dbname | database name | yugabyte

### Step 3: Create table

> Execute an SQL statement like the DDL `CREATE TABLE ...` using the `Exec()` function on the `db`

Reviewer: ```suggestion
Execute an SQL statement such as DDL `CREATE TABLE ...` using the `Exec()` function on the `db`
```

instance.

The CREATE DDL statement:

```sql
CREATE TABLE employee (id int PRIMARY KEY, name varchar, age int, language varchar)
```

Code snippet:

```go
var createStmt = `CREATE TABLE employee (id int PRIMARY KEY,
                                         name varchar,
                                         age int,
                                         language varchar)`;
if _, err := db.Exec(createStmt); err != nil {
    log.Fatal(err)
}
```

> The `db.Exec()` function also returns an `error` object which, if not `nil`, needs to handled in

Reviewer: ```suggestion
The `db.Exec()` function also returns an `error` object which, if not `nil`, needs to be handled in
```

All the functions or structs required for working with YugabyteDB database are part of `sql` package.

Use the `sql.Open()` function for getting connection object for the YugabyteDB database which can be
used for performing DDLs and DMLs against the database.

The connection details can be specified either as string params or via an url in the format given below:

```go
postgresql://username:password@hostname:port/database
```

Code snippet for connecting to YugabyteDB:

```go
psqlInfo := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s",
                        host, port, user, password, dbname)
// Other connection configs are read from the standard environment variables:
// PGSSLMODE, PGSSLROOTCERT, and so on.
db, err := sql.Open("postgres", psqlInfo)
defer db.Close()
if err != nil {
    log.Fatal(err)
}
```

| Parameter | Description | Default |
| :---------- | :---------- | :------ |
| host  | hostname of the YugabyteDB instance | localhost
| port |  Listen port for YSQL | 5433
| user | user for connecting to the database | yugabyte
| password | password for connecting to the database | yugabyte
| dbname | database name | yugabyte

> 

Author: Added SSL details.
